### PR TITLE
Gagaz Manqunazara

### PR DESCRIPTION
**Pull request by gazmanzara**
# Fancy Todo

### Day 1
- [x] Project init, install packages, and generating models
- [x] Get todos
- [x] Get todo by id
- [x] Post todo
- [x] Update todo
- [x] Update todo status
- [x] Delete todo
- [x] Documentation

### Day 2
- [x] Create model and association
- [x] Error handling
- [x] Post register
- [x] Post login
- [x] Jwt
- [x] Auth
- [x] Third party API

### Day 3
- [x] Jquery
- [x] Oauth

### Last
- [ ] Last documentation

### CODE REVIEW

File: server/controllers/todosController.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+            return res.status(200).json(data);
+        } catch (err) {
+            return res.status(500).json('Server Error');
+        }
+    }
+
+    static async postTodos (req, res) {
+        try {
+            let data = await Todo.create(req.body);
```
Comment (mhmmd-ysf) [inline server/controllers/todosController.js:16]: Jangan langsung lempar req.body

File: server/controllers/todosController.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            return res.status(201).json(data);
+        } catch (err) {
+            if (err.errors) {
+                let msg = [];
+
+                await err.errors.forEach(e => {
+                    msg.push(e.message);
+                });
+
+                return res.status(404).json({ name: 'Validation Error', msg: msg });
```
Comment (mhmmd-ysf) [inline server/controllers/todosController.js:35]: ingat 404 itu untuk `Not Found`, kalo ini jatohnya `Bad Request` karena salah input dari client, jadi harusnya status 400

File: server/models/todo.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+          msg: 'Status required!!!'
+        },
+        validFormat (value) {
+          if (value != 'unfinished' && value != 'finished') {
+            throw new Error('Status must be finished or unfinished!!!');
+          }
+        }
+      }
+    },
+    dueDate: {
```
Comment (mhmmd-ysf) [inline server/models/todo.js:39]: perhatikan nama properti yang diminta

**Review by mhmmd-ysf — COMMENTED** [review]
Overall udah oke

File: server/middlewares/auth.js [hunk 4]
```diff
@@ -1,11 +1,23 @@
...
+            if (data) {
+                next();
+            } else {
+                res.status(401).json({ msg: 'You must logged in!!!' });    
+            }
+        } else {
+            res.status(401).json({ msg: 'You must logged in!!!' });
+        }
+    } catch (err) {
+        res.status(500).json('Server Error');
```
Comment (mhmmd-ysf) [inline server/middlewares/auth.js:20]: errornya belom next ya

**Review by mhmmd-ysf — COMMENTED** [review]
Overall udah ok

File: server/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+node_modules
+.env
```
Comment (puppy-26) [inline server/.gitignore:2]: SECRET_KEY=thiismysecretkey
GOOGLE_AUTH_KEY=150552715425-vcgoiac96ndsbc205q919126gpip66ct.apps.googleusercontent.com
API_KEY=3567309decc72d4a1058e3e75bd58c10

**Comment by mhmmd-ysf** [comment]
Good Job!

**Pull request closed without merging**